File: client/world.py
from kivy.lang import Builder
from kivy.properties import ListProperty, NumericProperty, StringProperty
from kivy.uix.widget import Widget
from kivy.vector import Vector
from kivy.animation import Animation
from kivy.core.window import Window
from kivy.logger import Logger as log
from mediator import Mediator


class World(object):

    def __init__(self, parent, username):
        Builder.load_file('world.kv', rulesonly=True)
        self._widget = w = Widget()
        self._background_layer = background = Widget()
        main_pos = Window.width / 2, Window.height / 2
        self._main_layer = main = WorldLayer(pos=main_pos)
        self._uid = username
        self._objects = {}
        w.add_widget(background, 0)
        w.add_widget(main, 1)


        Mediator.subscribe(self, 'dispatch', 'recv.area')
        Mediator.subscribe(self, 'move_to', 'world.move_to')
        Mediator.subscribe(self, 'focus', 'world.focus')
        parent.add_widget(self._widget, 0)

    def deactivate(self):
        Mediator.unsubscribe_object(self)
        self._widget.parent.remove_widget(self._widget)

    def dispatch(self, cmd, body):
        getattr(self, "handle_" + cmd)(body)

    def _add_object(self, ident, data=None, index=0):
        tag = data['tag']
        if tag in ("User",):
            obj = Unit()
        else:
            raise ValueError("Unknown type '{}'".format(type))
        self._objects[ident] = obj
        self._main_layer.add_widget(obj, index)
        if tag in ("User",):
            obj.rel_x, obj.rel_y = data['pos']
            set_rel_pos(self.world_pos, obj)
            #obj.angle = data['angle']
        if tag == "User":
            obj.name = data['name']

    def _remove_object(self, ident):
        obj = self._objects.pop(ident)
        self._widget.remove_widget(obj)

    def handle_init(self, data):
        self._background_layer.add_widget(Background(x=0, y=0))

    @property
    def world_pos(self):
        return self._main_layer.pos

    def objects_info(self, objects):
        for data in objects:
            ident = data['id']
            if ident in self._objects:
                continue
            self._add_object(ident, data)

    def handle_tick(self, data):
        objects = self._objects
        unknown_objects = set()
        idents = set()
        pos = self.world_pos
        for value in data['objects']:
            ident = value['id']
            idents.add(ident)
            if ident not in objects:
                unknown_objects.add(ident)
                continue
            obj = objects[ident]
            obj.rel_x, obj.rel_y = value['pos']
            set_rel_pos(pos, obj)
        if unknown_objects:
            Mediator.publish("request",
                             "area.get_objects_info",
                             list(unknown_objects),
                             (self, 'objects_info'))
        for ident in set(objects) - idents:
            self._remove_object(ident)

    def focus(self, x, y):
        #TODO: use Vector
        layer = self._main_layer
        layer.x -= x - Window.width / 2
        layer.y -= y - Window.height / 2

    def move_to(self, x, y):
        wx, wy = self.world_pos
        Mediator.publish("send", "area.move_to", [x - wx, y - wy])


class Background(Widget):
    pass

class WorldLayer(Widget):

    def on_pos(self, _, pos):
        for ch in self.children:
            set_rel_pos(pos, ch)


class WorldObject(Widget):
    rel_x = NumericProperty(0)
    rel_y = NumericProperty(0)


class UnitImage(Widget):
    angle = NumericProperty(0)


class Unit(WorldObject):
    name = StringProperty()
    angle = NumericProperty(0)


def set_rel_pos((x, y), obj):
    obj.center_x = x + obj.rel_x
    obj.center_y = y + obj.rel_y

File: client/js/world.js
var FPS = 30;

function World(viewportEl, connection, user, area, ui) {
    this.viewportEl = viewportEl;
    this.connection = connection;
    this.user = user;
    this.area = area;
    this.ui = ui;
    this.serverTimeDiff = 0;
    this.animationLoopId = null;
    this.animationId = null;
    this.camera = new Camera();
    this.viewport = new ViewPort({el: viewportEl, model: this.camera});
    this.backgroundLayer = null;
    this.objectLayer = null;
    this.objectModels = {};
    this.appearanceReasons = {}; //FIXME: remove old events
    this.disappearanceReasons = {};
    this.firstEnter = true;
    this.documentFragment = null;
    _.bindAll(this, "animationLoopStep", "animationCallback");
    this.setupLayers();
    this.listenToConnection();
    this.listenToUI();
    this.requestAnimation();
}
World.prototype = Object.create(Backbone.Events);
_.extend(World.prototype, {
    constructor: World,
    setupLayers: function () {
        this.backgroundLayer = new Background({
            parallaxIndex: 0,
            model: this.camera,
            area: this.area,
            ui: this.ui
        });
        this.midgroundLayer = new Midground({
            parallaxIndex: .3,
            model: this.camera,
        });
        this.objectLayer = new Layer({parallaxIndex: 1, model: this.camera});
        this.backgroundLayer.$el.appendTo(this.viewportEl);
        this.midgroundLayer.$el.appendTo(this.viewportEl);
        this.objectLayer.$el.appendTo(this.viewportEl);
        this.listenTo(this.backgroundLayer, "move-to", this.moveTo);
    },
    listenToConnection: function () {
        this.listenTo(this.connection, "area", this.dispatch);
    },
    dispatch: function (data) {
        var handler = $.camelCase("handle-" + data.cmd);
        this[handler](data.body);
    },
    handleInit: function (data) {
        this.setServerTime(data.timestamp);
        _.each(_.keys(this.objectModels), this.removeObject, this);
        this.area.set({areaId: data.areaId, background: data.areaId + ".jpg"});
    },
    objectsInfo: function (objects) {
        var objectModels = this.objectModels;
        this.createDocumentFragment();
        _.each(objects, function (data) {
            if (_.has(objectModels, data.id)) return;
            this.addObject(data);
        }, this);
        this.applyDocumentFragment(this.objectLayer);
    },
    handleTick: function (data) {
        if (this.area.get("areaId") !== data.areaId) return;
        if (data.timestamp > this.getServerTime()) {
            this.setServerTime(data.timestamp);
            console.log("Server time updated");
        }
        var objectModels = this.objectModels;
        //TODO: use sets to improve time complexity
        var idents = [];
        var unknownIdents = [];
        var excessIdents = [];
        _.each(data.events, function (ev) {
            if (ev.tag === "Appearance") {
                this.appearanceReasons[ev.ident] = ev.aReason;
            }
            if (ev.tag === "Disappearance") {
                this.disappearanceReasons[ev.ident] = ev.dReason;
            }
        }, this);
        _.each(data.objects, function (value) {
            if (_.has(this.disappearanceReasons, value.id)) return;
            idents.push(value.id);
            if (!_.has(objectModels, value.id)) {
                unknownIdents.push(value.id);
            } else {
                objectModels[value.id].set(value);
            }
        }, this);
        if (!_.isEmpty(unknownIdents)) {
            this.connection.request(
                "area.get-objects-info",
                unknownIdents,
                this.objectsInfo,
                this
            );
        }
        excessIdents = _.difference(_.keys(objectModels), idents);
        _.each(excessIdents, this.removeObject, this);
        this.disappearanceReasons = [];
        this.animationStep();
    },
    getTime: function () {
        return performance.now();
    },
    setServerTime: function (serverTimestamp) {
        this.serverTimeDiff = this.getTime() - serverTimestamp;
    },
    getServerTime: function () {
        return this.getTime() - this.serverTimeDiff;
    },
    createDocumentFragment: function () {
        if (this.documentFragment !== null) return;
        this.documentFragment = $(document.createDocumentFragment());
    },
    applyDocumentFragment: function (layer) {
        if (this.documentFragment === null) return;
        this.documentFragment.appendTo(layer.$el);
        this.documentFragment = null;
    },
    addObject: function (data) {
        var reason = this.appearanceReasons[data.id];
        var model;
        var view;
        delete this.appearanceReasons[data.id];
        switch (data.tag) {
            case "User":
                if (!reason && data.id === this.user.get("userId")) {
                    reason = this.firstEnter ? "LogIn" : "Entry";
                }
                model = new Unit(data);
                view = new UserView({model: model, reason: reason});
                break;
            default:
                throw "Unknown type " + data.tag;
        }
        this.objectModels[data.id] = model;
        view.render().appendTo(this.documentFragment);
    },
    removeObject: function (ident) {
        var reason = this.disappearanceReasons[ident];
        this.objectModels[ident].trigger("destroy-view", reason);
        delete this.objectModels[ident];
    },
    moveTo: function (pos) {
        this.connection.send("area.move-to", pos.toArray());
    },
    listenToUI: function () {
        this.listenTo(this.ui, "focus-to-myself", this.showMyself);
        this.listenTo(this.ui, "ignite", this.ignite);
        this.listenTo(this.ui, "enter-area", this.enterArea);
    },
    showMyself: function () {
        var self = this.objectModels[this.user.get("userId")];
        if (self === undefined) return;
        this.camera.focusTo(Victor.fromArray(self.get("pos")));
    },
    ignite: function () {
        this.connection.send("area.ignite", 10);
    },
    enterArea: function (areaId) {
        this.firstEnter = false;
        var self = this.objectModels[this.user.get("userId")];
        self.trigger("destroy-view", "Exit");
        this.connection.send("area.enter-area", areaId);
    },
    requestAnimation: function () {
        this.animationLoopId = setTimeout(this.animationLoopStep, 1000 / FPS);
    },
    animationLoopStep: function () {
        this.animationId = requestAnimationFrame(this.animationCallback);
    },
    animationCallback: function () {
        this.requestAnimation();
        this.animationStep();
    },
    stopAnimationLoop: function () {
        clearTimeout(this.animationLoopId);
        cancelAnimationFrame(this.animationId);
    },
    animationStep: function () {
        var timestamp = this.getServerTime();
        _.each(_.values(this.objectModels), function (model) {
            if (model.applyActions !== undefined) {
                model.applyActions(timestamp);
            }
        }, this);
    }
});


